File: C10/countsubgrids.py
##O(n**3/N)
import time
def count(k): #in c++ use __builtin_popcount(k) runs in O(1)
    c=0
    for i in range(N):
        if k&1<<i:
            c+=1
    return c
start=time.time()
n=int(input())
g=[]
N=64
c=n//N+1
if n%N==0:
    c-=1
for i in range(n):
    s=input()
    g.append([])
    for j in range(c):
        k=0
        _from=j*N
        _to=min((j+1)*N,n)
        for i in range(_from,_to): 
            if s[i]=='1':
                k+=1<<i%N
        g[-1].append(k)

ans=0
for a in range(n-1):
    for b in range(a+1,n):
        cnt=0
        for i in range(c):
            cnt+=count(g[a][i]&g[b][i])
        ans+=(cnt-1)*cnt//2
end=time.time()
print(ans)
print(end-start)

File: C10/countsubgrids.cpp
#include <bits/stdc++.h>
#define ll long long
#define endl "\n"
using namespace std;

ll n;
int main(){
    cin >> n;
    ll N=64;
    ll c=n/N+1;
    if(n%N==0){
        c--;
    }
    vector<vector<ll>> g(n);
    for(int i=0;i<n;i++){
        string s;
        cin >> s;
        for(int j=0;j<c;j++){
            ll k=0;
            int from=j*N;
            int to=min((j+1)*N,n);
            for(int z=from;z<to;z++){
                if(s[z]=='1'){
                    k+=1ll<<z%N;
                }
            }
            g[i].push_back(k);    

        }
    }
    ll ans=0;
    for(ll a=0;a<n-1;a++){
        for(ll b=a+1;b<n;b++){
            ll cnt=0;
            for(ll i=0;i<c;i++){
                cnt+=__builtin_popcountll(g[a][i]&g[b][i]);
            }
            ans+=cnt*(cnt-1)/2ll;

        }
    }

    cout << ans << endl;
    return 0;
}